subprojects/testing-junit-platform/src/main/java/org/gradle/api/internal/tasks/testing/junitplatform/JUnitPlatformTestExecutionListener.java

```diff
@@ -60,119 +71,131 @@ public void testPlanExecutionFinished(TestPlan testPlan) {
-        return identifier.isTest() && !isVintageDynamicTestClass(identifier);
     }
 
     @Override
     public void executionFinished(TestIdentifier testIdentifier, TestExecutionResult testExecutionResult) {
-        if (testFailedBeforeTestClassStart(testIdentifier, testExecutionResult)) {
-            executionFailedBeforeTestClassStart(testIdentifier, testExecutionResult);
-        } else {
-            executionFinishedAfterTestClassStart(testIdentifier, testExecutionResult);
+        if (testExecutionResult.getStatus() == ABORTED) {
+            reportSkipped(testIdentifier);
+            return;
+        }
+        if (testExecutionResult.getStatus() == FAILED) {
+            reportStartedUnlessAlreadyStarted(testIdentifier);
+            Throwable failure = testExecutionResult.getThrowable().orElseGet(() -> new AssertionError("test failed but did not report an exception"));
+            if (testIdentifier.isTest()) {
+                resultProcessor.failure(getId(testIdentifier), failure);
+            } else {
+                TestDescriptorInternal syntheticTestDescriptor = createSyntheticTestDescriptorForTestClass(testIdentifier);
+                resultProcessor.started(syntheticTestDescriptor, startEvent(getId(testIdentifier)));
+                resultProcessor.failure(syntheticTestDescriptor.getId(), failure);
+                resultProcessor.completed(syntheticTestDescriptor.getId(), completeEvent());
+            }
+        }
+        if (wasStarted(testIdentifier)) {
+            resultProcessor.completed(getId(testIdentifier), completeEvent());
         }
     }
 
-    private void executionFailedBeforeTestClassStart(TestIdentifier testIdentifier, TestExecutionResult testExecutionResult) {
-        reportTestClassStarted(testIdentifier);
-        reportTestClassFinished(testIdentifier, testExecutionResult.getThrowable().get());
-    }
-
-    private void executionFinishedAfterTestClassStart(TestIdentifier testIdentifier, TestExecutionResult testExecutionResult) {
-        // Generally, there're 2 kinds of identifier:
-        // 1. A container (test engine/class/repeated tests). It is not tracked unless it fails/aborts.
-        // 2. A test "leaf" method. It's always tracked.
-        if (isFailedContainer(testIdentifier, testExecutionResult)) {
-            // only leaf methods triggered start events previously
-            // so here we need to add the missing start events
-            adapter.testStarted(testIdentifier.getUniqueId(), getDescriptor(testIdentifier));
-            testFinished(testIdentifier, testExecutionResult);
-        } else if (isLeafTest(testIdentifier)) {
-            testFinished(testIdentifier, testExecutionResult);
+    private void reportStartedUnlessAlreadyStarted(TestIdentifier testIdentifier) {
+        // guard against edge cases (e.g. JUnit 4 classes with custom runners that report the class as ignored after reporting it as started)
+        if (!wasStarted(testIdentifier)) {
```

Comment:
❌ Isn't there a race condition here, where two threads can check "wasStarted" and then both decide to call "started()"?